File: Core/src/main/java/me/mrkirby153/kcutils/event/UpdateEventHandler.java
package me.mrkirby153.kcutils.event;

import me.mrkirby153.kcutils.Module;
import org.bukkit.plugin.java.JavaPlugin;

public class UpdateEventHandler extends Module<JavaPlugin> implements Runnable {

    private int ticks = 0;

    public UpdateEventHandler(JavaPlugin plugin) {
        super("Update Event", "1.0", plugin);
    }

    @Override
    public void run() {
        // Prevent some sort of overflow if it's running too long. That could be bad
        if(++ticks >= Integer.MAX_VALUE){
            ticks = 1;
        }
        for(UpdateType t : UpdateType.values()){
            if(t.getUpdateTime() % ticks == 0){
                getPlugin().getServer().getPluginManager().callEvent(new UpdateEvent(t));
            }
        }
    }

    @Override
    protected void init() {
        scheduleRepeating(this, 0L, 1L);
    }
}

File: Bukkit/src/main/kotlin/me/mrkirby153/kcutils/event/UpdateEventHandler.kt
package me.mrkirby153.kcutils.event

import me.mrkirby153.kcutils.Module
import org.bukkit.plugin.java.JavaPlugin

/**
 * A module handling the firing of [UpdateEvents][UpdateEvent]
 *
 * @param plugin The owning plugin
 */
class UpdateEventHandler(plugin: JavaPlugin) : Module<JavaPlugin>("Update Event", plugin), Runnable {

    private var ticks = 0

    override fun run() {
        // Prevent some sort of overflow if it's running too long. That could be bad
        if (++ticks >= Integer.MAX_VALUE) {
            ticks = 1
        }
        UpdateType.values()
                .filter { it.elapsed(ticks) }
                .forEach { plugin.server.pluginManager.callEvent(UpdateEvent(it)) }
    }

    override fun init() {
        scheduleRepeating(0L, 1L, this)
    }
}
